### PR TITLE
use optional index in multivalued index

### LINKED ISSUE #2431: Fix inefficiency on multivalued but sparse column.
When dealing with multivalued column but sparse (most document have 0 values), tantivy is currently very inefficient.

We need to build an `start_offsets` index that is a column as long as the number of documents.

In particular, in dynamic mode, if a user is filling quickwit with documents with a dynamic key. 
`{"cart": {user_123123: ["item1", "item30"]}`
We could end up creating a lot of multivalued columns.

The phenomenon could be stress full on indexing.

On merge, we would have to extend these `start_offsets`, and sometime lift an efficient, sparse optional column to a multivalue.

This problem has been observed on Airmail.

# Possible solution

One possible solution would be to shield the multivalued column by an optional column marking the rows that have at least one document.

That solution is reasonable to implement and has being explored as a duct tape hack in https://github.com/quickwit-oss/tantivy/pull/2429.

Main cons are:
- multivalued columns will get more inefficient as every single call will have to go through the optional index access.
- when facing dense multivalued columns, the index will get 1 bit per document larger per such column.
We can add FULL blocks or BIZARRO blocks to the optional index to fix this, but the extra branching could have some unexpected effected on scalar operations microoptimizations.

### PR DESCRIPTION
**Pull request by PSeitz**
For mostly empty multivalued indices there was a large overhead during
creation when iterating all docids. This is alleviated by placing an
optional index in the multivalued index to mark documents that have values.

There's some performance overhead when accessing values in a multivalued
index. The accessing cost is now optional index + multivalue index. The
sparse codec performs relatively bad with the binary_search when accessing
data. This is reflected in the benchmarks below.

This changes the format of columnar to v2, but code is added to handle the v1
formats. This is done by passing down the global version flag to the index. We could have the version in each column, this would make version handling slightly less awkward. But it adds some extra bytes, which may add up, if there are many columns.

```bash
     Running benches/bench_access.rs (/home/pascal/Development/tantivy/optional_multivalues/target/release/deps/bench_access-ea323c028db88db4)
multi sparse 1/13
access_values_for_doc        Avg: 42.8946ms (+241.80%)    Median: 42.8869ms (+244.10%)    [42.7484ms .. 43.1074ms]
access_first_vals            Avg: 42.8022ms (+421.93%)    Median: 42.7553ms (+439.84%)    [42.6794ms .. 43.7404ms]
multi 2x
access_values_for_doc        Avg: 31.1244ms (+24.17%)    Median: 30.8339ms (+23.46%)    [30.7192ms .. 33.6059ms]
access_first_vals            Avg: 24.3070ms (+70.92%)    Median: 24.0966ms (+70.18%)    [23.9328ms .. 26.4851ms]
sparse 1/13
access_values_for_doc        Avg: 42.2490ms (+0.61%)    Median: 42.2346ms (+2.28%)    [41.8988ms .. 43.7821ms]
access_first_vals            Avg: 43.6272ms (+0.23%)    Median: 43.6197ms (+1.78%)    [43.4920ms .. 43.9009ms]
dense 1/12
access_values_for_doc        Avg: 8.6184ms (+23.18%)    Median: 8.6126ms (+23.78%)    [8.5843ms .. 8.7527ms]
access_first_vals            Avg: 6.8112ms (+4.47%)     Median: 6.8002ms (+4.55%)     [6.7887ms .. 6.8991ms]
full
access_values_for_doc        Avg: 9.4073ms (-5.09%)    Median: 9.4023ms (-2.23%)    [9.3694ms .. 9.4568ms]
access_first_vals            Avg: 4.9531ms (+6.24%)    Median: 4.9502ms (+7.85%)    [4.9423ms .. 4.9718ms]
```

```bash
     Running benches/bench_merge.rs (/home/pascal/Development/tantivy/optional_multivalues/target/release/deps/bench_merge-475697dfceb3639f)
merge_multi 2x_and_multi 2x                          Avg: 20.2280ms (+34.33%)    Median: 20.1829ms (+35.33%)    [19.9933ms .. 20.8806ms]
merge_multi sparse 1/13_and_multi sparse 1/13        Avg: 0.8961ms (-78.04%)     Median: 0.8943ms (-77.61%)     [0.8899ms .. 0.9272ms]
merge_dense 1/12_and_dense 1/12                      Avg: 0.6619ms (-1.26%)      Median: 0.6616ms (+2.20%)      [0.6473ms .. 0.6837ms]
merge_sparse 1/13_and_sparse 1/13                    Avg: 0.5508ms (-0.85%)      Median: 0.5508ms (+2.80%)      [0.5420ms .. 0.5634ms]
merge_sparse 1/13_and_dense 1/12                     Avg: 0.6046ms (-4.64%)      Median: 0.6038ms (+2.80%)      [0.5939ms .. 0.6296ms]
merge_multi sparse 1/13_and_dense 1/12               Avg: 0.9111ms (-83.48%)     Median: 0.9063ms (-83.50%)     [0.9047ms .. 0.9663ms]
merge_multi sparse 1/13_and_sparse 1/13              Avg: 0.8451ms (-89.49%)     Median: 0.8428ms (-89.43%)     [0.8411ms .. 0.8563ms]
merge_multi 2x_and_dense 1/12                        Avg: 10.6624ms (-4.82%)     Median: 10.6568ms (-4.49%)     [10.5738ms .. 10.8353ms]
merge_multi 2x_and_sparse 1/13                       Avg: 10.6336ms (-22.95%)    Median: 10.5925ms (-22.33%)    [10.5149ms .. 11.5657ms]
```

Based on: https://github.com/quickwit-oss/tantivy/pull/2429

Closes https://github.com/quickwit-oss/tantivy/issues/2431

### CODE REVIEW

File: columnar/src/column_index/optional_index/set_block/sparse.rs [hunk 1]
```diff
@@ -109,3 +109,24 @@ impl<'a> SparseBlock<'a> {
...
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::column_index::optional_index::set_block::DenseBlockCodec;
+
+    #[test]
+    fn test_simple_translate_idx_to_value_idx_dense() {
+        let mut buffer = Vec::new();
+        DenseBlockCodec::serialize([1, 10].iter().copied(), &mut buffer).unwrap();
```
Comment (fulmicoton) [inline columnar/src/column_index/optional_index/set_block/sparse.rs]: Is this the right file for this test? I would have expected it in dense.rs

Comment (PSeitz) [reply columnar/src/column_index/optional_index/set_block/sparse.rs]: ah yes, I already moved it, but forgot to delete it

File: columnar/src/columnar/format_version.rs [hunk 2]
```diff
@@ -62,7 +65,7 @@ mod tests {
     #[test]
     fn test_footer_dserialization() {
```
Comment (fulmicoton) [inline columnar/src/columnar/format_version.rs]:
```suggestion
    fn test_footer_deserialization() {
```

File: columnar/src/column_index/mod.rs [hunk 3]
```diff
@@ -131,15 +134,41 @@ impl ColumnIndex {
...
+                    // * doc_id_range.start may exist but doc_id_range.end may not exist
+                    // * doc_id_range.start and doc_id_range.end may exist
+                    // => rank on doc_id_range.end will give use the next value, which matches
+                    // how the `start_index_column` works, so we get the value start of the next
+                    // docid which we use to create the exclusive range.
+                    //
+                    let rank = index.optional_index.rank(doc_id_range.start);
+                    let row_start = index.start_index_column.get_val(rank);
+                    let rank = index.optional_index.rank(doc_id_range.end);
+                    let row_end = index.start_index_column.get_val(rank);
```
Comment (fulmicoton) [inline columnar/src/column_index/mod.rs]:
```suggestion
                    let rank_start = index.optional_index.rank(doc_id_range.start);
                    let row_start = index.start_index_column.get_val(rank_start);
                    let rank_end = index.optional_index.rank(doc_id_range.end);
                    let row_end = index.start_index_column.get_val(rank_end);
```

**Review by fulmicoton — APPROVED** [review]

**Pull request merged**